feat: add C implementation for `stats/base/dists/triangular/logcdf`

0PrashantYadav0 opened the pull request:
Resolves #3814

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for @stdlib/stats/base/dists/triangular/logcdf along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3814

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
refactor: update `stats/base/dnanmeanpn` native addon from C++ to C

PR-URL: https://github.com/stdlib-js/stdlib/pull/4110
Reviewed-by: Athan Reines <kgryte@gmail.com>

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

0PrashantYadav0 committed:
Merge branch 'develop' of https://github.com/0PrashantYadav0/stdlib into develop

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
Merge branch 'develop' of https://github.com/0PrashantYadav0/stdlib into develop

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
Merge branch 'stdlib-js:develop' into develop

0PrashantYadav0 committed:
feat: add C implementation for triangular logcdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
fix: changes in comment and test inputs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/README.md:
```diff
@@ -158,6 +158,109 @@ for ( i = 0; i < 25; i++ ) {
+-   **x**: `[in] double` input value.
+-   **a**: `[in] double` lower limit
+-   **b**: `[in] double` upper limit
+-   **c**: `[in] double` mode
```
```suggestion
-   **a**: `[in] double` lower limit.
-   **b**: `[in] double` upper limit.
-   **c**: `[in] double` mode.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/README.md:
```diff
@@ -158,6 +158,109 @@ for ( i = 0; i < 25; i++ ) {
+        x = random_uniform( 0.0, 30.0 );
+        a = random_uniform( 0.0, 10.0 );
+        b = random_uniform( a, 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+        c = a + random_uniform( 0.0, b - a );
```
```suggestion
        c = random_uniform( a, b );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+		x[ i ] = random_uniform( 0.0, 30.0 );
+		a[ i ] = random_uniform( 0.0, 10.0 );
+		b[ i ] = random_uniform( a[ i ], 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+		c[ i ] = a[ i ] + random_uniform( 0.0, b[ i ] - a[ i ] );
```
```suggestion
		c[ i ] = random_uniform( a[ i ], b[ i ] );
```

Planeshifter commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

0PrashantYadav0 commented:
@Planeshifter Please review this pr if possible

0PrashantYadav0 commented:
/stdlib update-copyright-year

0PrashantYadav0 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

0PrashantYadav0 committed:
chore: resolve lint error and minor changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
@Planeshifter Please review and merge this pr

0PrashantYadav0 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into triangular/logcdf

0PrashantYadav0 committed:
fix: changes in benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
Merge branch 'develop' into triangular/logcdf

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/README.md:
```diff
@@ -158,6 +158,105 @@ for ( i = 0; i < 25; i++ ) {
+    for ( i = 0; i < 25; i++ ) {
+        x = random_uniform( 0.0, 30.0 );
+        a = random_uniform( 0.0, 10.0 );
+        b = random_uniform( a, 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
        b = random_uniform( a + STDLIB_CONSTANT_FLOAT64_EPS, 40.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/benchmark/benchmark.js:
```diff
@@ -34,17 +35,26 @@ bench( pkg, function benchmark( b ) {
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( 0.0, 30.0 );
+		min[ i ] = uniform( 0.0, 10.0 );
+		max[ i ] = min[ i ] + ( uniform( 0.0, 40.0 ) ) + EPS;
+		mode[ i ] = min[ i ] + ( ( max[ i ] - min[ i ] ) * uniform( 0.0, 1.0 ) ); // eslint-disable-line max-len
```
```suggestion
		max[ i ] = uniform( min[ i ] + EPS, min[ i ] + 40.0 ) );
		mode[ i ] = uniform( min[ i ], max[ i ] );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/examples/c/example.c:
```diff
@@ -0,0 +1,46 @@
+	for ( i = 0; i < 25; i++ ) {
+		x = random_uniform( 0.0, 30.0 );
+		a = random_uniform( 0.0, 10.0 );
+		b = random_uniform( a, 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
+		c = a + random_uniform( 0.0, b - a );
```
```suggestion
		b = random_uniform( a + STDLIB_CONSTANT_FLOAT64_EPS, 40.0 );
		c = random_uniform( a, b );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/src/main.c:
```diff
@@ -0,0 +1,59 @@
+* @return    evaluated logCDF
+*
+* @example
+* double y = stdlib_base_dists_geometric_logcdf( 0.5, -1.0, 1.0, 0.0 );
```
```suggestion
* double y = stdlib_base_dists_triangular_logcdf( 0.5, -1.0, 1.0, 0.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/manifest.json:
```diff
@@ -0,0 +1,90 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/eps",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/ninf"
```
Duplicate entry:

```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/constants/float64/eps",
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/special/pow"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/manifest.json:
```diff
@@ -0,0 +1,90 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/eps",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/ninf"
```
Duplicate entry:

```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/constants/float64/eps",
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/special/pow"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/manifest.json:
```diff
@@ -0,0 +1,90 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/quaternary",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/math/base/special/ln",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/ninf"
```
Duplicate entry:

```suggestion
        "@stdlib/math/base/napi/quaternary",
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/constants/float64/ninf",
        "@stdlib/math/base/special/ln",
        "@stdlib/math/base/special/pow"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/include/stdlib/stats/base/dists/triangular/logcdf.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the natural logarithm of the cumulative distribution function for the triangular distribution with parameters a, b, and c.
```
```suggestion
* Evaluates the natural logarithm of the cumulative distribution function for the triangular distribution with lower limit `a`, upper limit `b`, and mode `c` at a value `x`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/src/main.c:
```diff
@@ -0,0 +1,59 @@
+#include "stdlib/math/base/special/pow.h"
+
+/**
+* Evaluates the natural logarithm of the cumulative distribution function (CDF) for a triangular distribution with lower limit `a` and upper limit `b` and mode `c` at a value `x`.
```
```suggestion
* Evaluates the natural logarithm of the cumulative distribution function (CDF) for a triangular distribution with lower limit `a`, upper limit `b`, and mode `c` at a value `x`.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,145 @@
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = random_uniform( 0.0, 30.0 );
+		a[ i ] = random_uniform( 0.0, 10.0 );
+		b[ i ] = random_uniform( a[ i ], 40.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		b[ i ] = random_uniform( a[ i ] + STDLIB_CONSTANT_FLOAT64_EPS, 40.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/triangular/logcdf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,77 @@
+	for ( i = 0; i < len; i++ ) {
+		x[ i ] = uniform( 0.0, 30.0 );
+		min[ i ] = uniform( 0.0, 10.0 );
+		max[ i ] = min[ i ] + ( uniform( 0.0, 40.0 ) ) + EPS;
+		mode[ i ] = min[ i ] + ( ( max[ i ] - min[ i ] ) * uniform( 0.0, 1.0 ) ); // eslint-disable-line max-len
```
```suggestion
		max[ i ] = uniform( min[ i ] + EPS, 40.0 );
		mode[ i ] = uniform( min[ i ], max[ i ] );
```

Planeshifter reviewed (changes requested):
Thanks, @0PrashantYadav0! 
Left some comments; once addressed, should be ready to land.

0PrashantYadav0 committed:
chore: changes as per code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
@Planeshifter Made all the changes please review and merge.

Planeshifter reviewed (approved):
LGTM!
